### ISSUE TITLE
GITHUB_TOKEN read-only breaks github action

### ISSUE THREAD

**Issue by chrisgilmerproj** [hmarr/auto-approve-action#183]
See #180 for some details on this issue.

March 1st, Github changed the `GITHUB_TOKEN` to be read-only for all workflows. [See the announcement dated Feb 19, 2020](https://github.blog/changelog/2021-02-19-github-actions-workflows-triggered-by-dependabot-prs-will-run-with-read-only-permissions/). This has to do with security vulnerabilities with using the `pull_request_target` trigger and scoping permissions. [See guide on Preventing Pwn Requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests). 

Unfortunately this breaks the integration as seen here:

<img width="416" alt="Screen Shot 2021-03-10 at 9 51 19 AM" src="https://user-images.githubusercontent.com/219296/110674182-7f85ad80-8186-11eb-9382-802611c83d16.png">

Changing the trigger to `pull_request` does not fix it because `write` permissions are needed to create a review of the PR.

While I don't have a solution for fixing this I have reconfigured my job to look like this per the security guidelines, primarily using labels:

```yml
name: auto approve
on:
  pull_request_target:
    types: [labeled]

jobs:
  auto-approve:
    runs-on: ubuntu-latest
    steps:
      - uses: hmarr/auto-approve-action@v2.0.0
        if: github.actor == 'dependabot[bot]' && contains(github.event.pull_request.labels.*.name, 'dependencies')
        with:
          github-token: "${{ secrets.GITHUB_TOKEN }}"
```

I want to thank @skjnldsv and @alexwilson for having provided the context I needed to dig into this.

**Comment by skjnldsv** [comment]
> primarily using labels:

Hey! What does labels change in this context? :)
Thanks for sharing!

**Comment by chrisgilmerproj** [comment]
@skjnldsv - from the security guide it says this:

> Since external users do not have the permission to assign labels, this effectively requires repository owners to manually review the changes first and is prone to human error.

I believe the idea is to ensure that if someone forks your repo and makes a PR back that they can't run a Github Action that includes code that can exfiltrate your secrets. I see this as an odd kind of hack but worth keeping in mind for this type of tool.

**Comment by tclass** [comment]
are there any ideas on how to make this work again? Can I host the workflow myself and then get a write token for example?

**Comment by jotoeri** [comment]
I'm not sure, if  @chrisgilmerproj did something different, but for me it seems to work with using `pull_request_target`. You just don't see the action on the implementing pull-request. (As your PR thinks it's done on base-branch, while base-branch thinks it's done on PR 😄)
For small extra security, i set the base-branch to be master, so people will not be able to create a PR onto another PR (but that is an extra gimmick to what was earlier).
See https://github.com/nextcloud/forms/pull/854 and https://github.com/nextcloud/forms/pull/855 for my considerations, if it really works in production we'll see probably next week. 🙈

**Comment by alexwilson** [comment]
In my action, which pairs nicely with this one, I moved to using the `labeled` event for `pull_request_target` and explicitly removed any checked-out code: https://github.com/alexwilson/enable-github-automerge-action/blob/main/.github/workflows/auto-merge-dependabot.yml

There's also a more practical example of this in a single workflow here: https://github.com/alexwilson/frontend/blob/master/.github/workflows/auto-approve.yml

Both, so far, are working as before - and while the communication from GitHub wasn't ideal it was a useful exercise because it encouraged me to think about 3rd party code having access to secrets.  @JakeChampion made an excellent recommendation to lock 3rd party actions to a specific commit hash since they are (in-theory) immutable which reduces risk even further.

**Comment by alexwilson** [comment]
The key thing to consider is: A lot of this depends on the `actor` firing the event.  If it's dependabot - both `pull_request` and the new `pull_request_target` with the default `created` & `edited` events will *not* allow read-access to secrets.

`labeled`, however, _does_ allow this - and only special cases (i.e. `dependabot` or other explicitly installed apps) are still able to label PRs despite not having full access to the repository.  So it is "safer".

Another change I made here - is checking for the creator of the pull-request, not the actor associated with an event, so that the action can be re-triggered & debugged manually.  If / when this action fails again, I'll be able to re-trigger the workflow and since the original creator was `dependabot` - it works as normal.
```yaml
    if: github.event.pull_request.user.login == 'dependabot[bot]'
```

**Comment by alexwilson** [comment]
Following all of this - [I added a disclaimer & adjusted the docs for the `enable-github-automerge` action](https://github.com/alexwilson/enable-github-automerge-action#2-usage), and highly recommend you do the same here @hmarr (or anyone else wishing to contribute to this excellent action) 😄

**Comment by skjnldsv** [comment]
> I believe the idea is to ensure that if someone forks your repo and makes a PR back that they can't run a Github Action that includes code that can exfiltrate your secrets. I see this as an odd kind of hack but worth keeping in mind for this type of tool.

Yeah, this doesn't really scale for us as we have a lot of external contributors at @nextcloud. This is our open contributing policy.

**Comment by jotoeri** [comment]
> default created & edited events will not allow read-access to secrets.

@alexwilson Do you have any source for this? I currently don't find it on the docs... 🤔 Just want to read a bit to know, what i'm doing there...

**Comment by skjnldsv** [comment]
But for me, even with `pull_request_target`, it doesn't work, I have a `action triggered outside of a pull_request` error :thinking:

**Comment by alexwilson** [comment]
@skjnldsv 
I tried to answer this above - you are fine with an open contributing policy - the primary restriction is you should probably be using `pull_request_target` for any workflows that need to write back to your repo _because_ pull-request-target operates in the base branch (i.e. 3rd party code cannot accidentally run!).

If you need to _also_ do something with the code itself (i.e. run tests) - `pull_request` actions run _without any secrets_ in the branch being PR'd.  For more complicated edge-cases (i.e. you want to provide a BrowserStack token to do some E2E testing) GitHub's recommendation is you invoke these workflows manually _after_ auditing the code.

@jotoeri 
The docs are awfully confusing - this is intel pieced together from a Github Support thread, many pages of docs, and the two blog posts GitHub have on this.  I'm feeding back in the support thread, so hopefully this will be clarified.

The intent is as you summarised however - so if you bear that in mind you should be okay!

**Comment by jotoeri** [comment]
> But for me, even with `pull_request_target`, it doesn't work, I have a `action triggered outside of a pull_request` error 🤔

@skjnldsv I saw this on the auto-merge-action, as it supports the `pull_request_target` only since a few days. If we bump `ahmadnassri/action-dependabot-auto-merge` to `@v2` it works. (or specific commit, as alex mentioned).

**Comment by hmarr** [comment]
Sorry for taking a while to get to this!

**I think switching from `pull_request` to `pull_request_target` should work.** I'm pretty sure that the read-only tokens introduced last month should only apply to Dependabot pull requests triggered by `pull_request` events. Switching to `pull_request_target` will run the workflow in the target branch (e.g. `main` rather than the PR's branch), but that's not a problem for this action as it just leaves a review and doesn't look at the code in the pull request.

I tried both, and confirmed that `pull_request_target` does work as expected:

#### `pull_request`
[This actions run](https://github.com/hmarr/auto-approve-test/runs/2116761899?check_suite_focus=true) was triggered using the `pull_request` event, and it failed because the token only had `read` access to pull requests.

<img width="903" alt="Screenshot 2021-03-15 at 18 54 36" src="https://user-images.githubusercontent.com/110275/111231743-b07c3d00-85e1-11eb-8e04-446874ce1da3.png">


#### `pull_request_target`
[This action](https://github.com/hmarr/auto-approve-test/runs/2116761884?check_suite_focus=true) was triggered using `pull_request_target` event. It worked as expected.

<img width="900" alt="Screenshot 2021-03-15 at 18 53 59" src="https://user-images.githubusercontent.com/110275/111231729-a65a3e80-85e1-11eb-8549-ea972573cfc8.png">

I'm adding some [more useful error messages](https://github.com/hmarr/auto-approve-action/blob/ce1c7f0f08b012f021130d2618caf6ca4d614e13/src/approve.ts#L36-L40), which should help folks catch this in the future.

**If you're seeing "Resource not accessible by integration" with a workflow file (in a public repo) that _already_ has been switched over to `pull_request_target`, can you point me to the actions run?** I don't think that should be happening, so if it is I'd like to investigate.

_Aside: I don't think the linked [Security Lab blog post](https://securitylab.github.com/research/github-actions-preventing-pwn-requests) is relevant. It raises a security issue associated with checking out potentially vulnerable code in the course of an Actions workflow, which this action doesn't do._

**Comment by chrisgilmerproj** [comment]
Thank you for helping with this ticket @hmarr . I believe you are correct that switching to `pull_request_target` fixes it. I'm going to close this and hopefully other folks will be able to use it for updating their own code.

**Comment by hmarr** [comment]
Glad to hear it's working!

I just published [v2.1.0](https://github.com/hmarr/auto-approve-action/releases/tag/v2.1.0) which includes better error handling. If you want to track v2 rather than pinning to a specific patch release you can also use `hmarr/auto-approve-action@v2` now.

**Comment by ReenigneArcher** [comment]
I've been using this action for about 2 years I think... just faced this on a new repository.

Had to manually change these settings within the repo.

![image](https://github.com/hmarr/auto-approve-action/assets/42013603/a70a5830-5e81-473b-a525-cfaa2cdd37b6)

I actually have this set as org defaults in my github org... but this repo was transferred to my org so it didn't pick up our defaults.

`pull_request_target` is not necessary!
